src/coreclr/jit/morph.cpp

```diff
@@ -323,6 +323,222 @@ GenTree* Compiler::fgMorphExpandCast(GenTreeCast* tree)
             }
         }
     }
+
+    // This if check needs to be changed to make sure we only
+    // block casts which are already Fixed UP.
+    do
+    {
+        if (!tree->gtOverflow() && varTypeIsFloating(srcType) && varTypeIsIntegral(dstType) && !varTypeIsSmall(dstType))
```

Comment:
I slightly worry that doing this in morph may hinder other optimizations (value numbering, constant folding, etc) and that it might be better to do in lowering.

I don't think its worth changing that for this PR, and its better to get it in working; but its probably something to look at long term.